2-WebApp-graph-user/2-6-BFF-Proxy/CallGraphBFF/Utils/CustomCookieAuthenticationEvents.cs

```diff
@@ -0,0 +1,49 @@
+﻿using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Web;
+using System.Net;
+
+internal class CustomCookieAuthenticationEvents : CookieAuthenticationEvents
+{
+    // Respond with 401 instead of redirect to IdP for unauthenticated attempts to access a secure route
+    public override Task RedirectToLogin(RedirectContext<CookieAuthenticationOptions> context)
+    {
+        context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+        return Task.CompletedTask;
+    }
+
+    public override Task RedirectToAccessDenied(RedirectContext<CookieAuthenticationOptions> context)
+    {
+        context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
+        return Task.CompletedTask;
+    }
+
+    // Reject authentication cookie if no tokens are found in the cache
+    public async override Task ValidatePrincipal(CookieValidatePrincipalContext context)
+    {
+        try
```

Comment:
that's a great experience. 
@jennyf19 : would solve https://github.com/AzureAD/microsoft-identity-web/issues/13  elegantly